### ISSUE TITLE
Make `operations.spec.ts` output more usable

### ISSUE THREAD

**Issue by anicolao** [cubing/cubing.js#206]
**To Reproduce**

1. Add a failing test to `operations.spec.ts` or make an existing test fail
2. Run `make test-spec`

Output is of the form:
```
$ make test-spec
npm run test-spec

> cubing@0.28.3 test-spec
> npx web-test-runner


src/cubing/alg/operation.spec.ts:

 ❌ operation > can append moves
      AssertionError: expected false to be true
      + expected - actual
      
      -false
      +true
      
      at o.<anonymous> (src/cubing/alg/operation.spec.ts:29:12)

Chrome: |██████████████████████████████| 21/21 test files | 154 passed, 1 failed

Code coverage: 60.85 %
View full coverage report at .temp/coverage/lcov-report/index.html

Finished running tests in 3s with 1 failed tests.

make: *** [test-spec] Error 1
```

**Expected behavior**

The `true` vs `false` really doesn't help the user tackle the issue. I propose that we make the test legible by expecting the string values to match which produces much nicer output. For example:
```
$ make test-spec
npm run test-spec

> cubing@0.28.3 test-spec
> npx web-test-runner


src/cubing/alg/operation.spec.ts:

 ❌ operation > can append moves experimentalAppendMove of R U R' + U2 => R U R U2
      AssertionError: expected 'R U R\' U2' to equal 'R U R U2'
      + expected - actual
      
      -R U R' U2
      +R U R U2
      
      at o.<anonymous> (src/cubing/alg/operation.spec.ts:12:9)

Chrome: |██████████████████████████████| 21/21 test files | 154 passed, 1 failed

Code coverage: 62.3 %
View full coverage report at .temp/coverage/lcov-report/index.html

Finished running tests in 3.1s with 1 failed tests.

make: *** [test-spec] Error 1
```

**Comment by anicolao** [comment]
Here is my proposed option 1 for converting these tests:

```
diff --git a/src/cubing/alg/operation.spec.ts b/src/cubing/alg/operation.spec.ts
index 15410d33..e612af8f 100644
--- a/src/cubing/alg/operation.spec.ts
+++ b/src/cubing/alg/operation.spec.ts
@@ -4,13 +4,24 @@ import { Alg } from "./Alg";
 import { experimentalAppendMove } from "./operation";
 import { Move } from "./alg-nodes";
 
+function testAppendMoveTransform({ test, start, move, result, options }) {
+  it(`${test} experimentalAppendMove of ${start} + ${move} => ${result}`, () =>
+    expect(
+      experimentalAppendMove(
+        new Alg(start), new Move(move), options).toString(),
+    ).to.equal(result)
+  );
+}
+
+
 describe("operation", () => {
+	testAppendMoveTransform({
+		test: "can append moves",
+		start: "R U R'",
+		move: "U2",
+		result: "R U R U2"
+	});
   it("can append moves", () => {
-    expect(
-      experimentalAppendMove(new Alg("R U R'"), new Move("U2")).isIdentical(
-        new Alg("R U R' U2"),
-      ),
-    ).to.be.true;
     expect(
       experimentalAppendMove(new Alg("R U R'"), new Move("R", -2)).isIdentical(
         new Alg("R U R' R2'"),
```

Pros: verbose helper function makes it very clear to see what each test consists of.
Cons: still a lot of code per test. 

Option 2 in the next comment.

**Comment by anicolao** [comment]
Here is Option 2. I prefer it because I find the tests themselves very legible, and it will make the tests short and readable both in the `spec` file and in case of any failures.

```
diff --git a/src/cubing/alg/operation.spec.ts b/src/cubing/alg/operation.spec.ts
index e612af8f..1a5e6166 100644
--- a/src/cubing/alg/operation.spec.ts
+++ b/src/cubing/alg/operation.spec.ts
@@ -13,26 +13,35 @@ function testAppendMoveTransform({ test, start, move, result, options }) {
   );
 }
 
+function tests({ test, options, tests }) {
+	tests.map(s => {
+		let parts = s.split(/[+=]/);
+		expect(parts.length).to.equal(3);
+		testAppendMoveTransform({
+			test,
+			start: parts[0].trim(),
+			move: parts[1].trim(),
+			result: parts[2].trim(),
+			options
+		})
+	});
+}
 
 describe("operation", () => {
 	testAppendMoveTransform({
 		test: "can append moves",
 		start: "R U R'",
 		move: "U2",
-		result: "R U R U2"
+		result: "R U R' U2"
+	});
+  tests({ 
+  	test: "can append moves",
+  	options: {},
+  	tests: [
+      "R U R' + R2' = R U R' R2'",
+      "R U R' + R = R U R' R",
+		],
 	});
-  it("can append moves", () => {
-    expect(
-      experimentalAppendMove(new Alg("R U R'"), new Move("R", -2)).isIdentical(
-        new Alg("R U R' R2'"),
-      ),
-    ).to.be.true;
-    expect(
-      experimentalAppendMove(new Alg("R U R'"), new Move("R")).isIdentical(
-        new Alg("R U R' R"),
-      ),
-    ).to.be.true;
-  });
 
   it("can coalesce appended moves", () => {
     expect(
```

**Comment by anicolao** [comment]
@lgarron @rokicki please comment on which of these two styles you'd prefer, or ask me not to make this change at all, and I will put up the appropriate pull request to fix the file.

I expect to be adding a large number of tests to this file with later patches for slice moves, wide moves, and changes to how `mod` and `coalesce` work, so I'd like it to be as concise and understandable as possible for both test creation and fixing.

**Comment by lgarron** [comment]
We previously had a helper: https://github.com/cubing/cubing.js/blob/190ee1d7aefe15fb0af8d092671d621ab11df1d3/src/cubing/alg/test/structure-equals.ts

I removed that helper while moving our tests to `chai`, but I believe `chai` also has a way to define custom comparison: https://www.chaijs.com/guide/helpers/
I'd strongly prefer to bring back such an implementation.

As for having lots of tests, I've learned it's best to avoid too much indirection in testing. However, your last comment looks a lot like table-driven testing. I think it makes sense to use for any bulk comparison testing, although I think the algs should be passed as separate arguments instead of using ` = `. If you're up for that, we can discuss more details in a PR.

**Comment by anicolao** [comment]
So you'd prefer something like the following?

```
tests({
    test: "slice moves",
    options: { wideMoves: true, sliceMoves: true, },
    tests: [
      { start: "R' R", move: "x", result: "R' R x" },
      { start: "L' R", move: "x'", result: "M" },
    ],
  });
```

**Comment by anicolao** [comment]
I don't know how to do the chai part because there is nowhere to import the types from. A different solution than `src/test/chai-workaround/index.ts` seems desirable to me and/or I don't understand the decisions underlying that workaround (which may also be the source of the difficulty removing mocha types in https://github.com/cubing/cubing.js/issues/209 which I also looked at briefly and gave up on).

**Comment by anicolao** [comment]
After much futzing around, I came up with this output:
```
$ make test-spec
npm run test-spec

> cubing@0.28.4 test-spec
> npx @web/test-runner --playwright


src/cubing/alg/operation.spec.ts:

 ❌ operation > can append moves
      AssertionError: expected 'R U R\' U2' to equal 'R U R\' U'
      + expected - actual
      
      -R U R' U2
      +R U R' U
      
      at Proxy.compareAlgs (src/cubing/alg/operation.spec.ts:14:28)
      at Proxy.f (node_modules/@esm-bundle/chai/node_modules/.pnpm/chai@4.3.4/node_modules/chai/chai.js:9164:32)
      at o.<anonymous> (src/cubing/alg/operation.spec.ts:27:9)

Chromium: |██████████████████████████████| 21/21 test files | 154 passed, 1 failed

Code coverage: 68.91 %
View full coverage report at .temp/coverage/lcov-report/index.html

Finished running tests in 3.6s with 1 failed tests.

make: *** [test-spec] Error 1
```

by doing this:

```
diff --git a/src/cubing/alg/operation.spec.ts b/src/cubing/alg/operation.spec.ts
index ee52a9d9..71ccbb4d 100644
--- a/src/cubing/alg/operation.spec.ts
+++ b/src/cubing/alg/operation.spec.ts
@@ -1,28 +1,30 @@
-import { expect } from "../../test/chai-workaround";
+import { expect, use } from "../../test/chai-workaround";
 
 import { Alg } from "./Alg";
 import { experimentalAppendMove } from "./operation";
 import { Move } from "./alg-nodes";
 
-Assertion.overwriteMethod('equal', function (_super) {
-  return function compareAlgs (other) {
-		var obj = this._obj;
-    if (obj && obj instanceof Alg) {
-      const actualString = obj.toString();
-      const expectedString = other.toString();
-      expect(actualString).to.equal(expectedString);
-      expect(obj.isIdentical(other));
-    } else {
-      _super.apply(this, arguments);
-    }
-  };
+use(function (chai, utils) {
+  chai.Assertion.overwriteMethod("equal", function (_super) {
+    return function compareAlgs(other) {
+      const obj = this._obj;
+      if (obj && obj instanceof Alg) {
+        const actualString = obj.toString();
+        const expectedString = other.toString();
+        expect(actualString).to.equal(expectedString);
+        expect(obj.isIdentical(other));
+      } else {
+        _super.apply(this, arguments);
+      }
+    };
+  });
 });
 
 describe("operation", () => {
   it("can append moves", () => {
-    expect(
-      experimentalAppendMove(new Alg("R U R'"), new Move("U2"))
-    ).to.equal(new Alg("R U R' U2"));
+    expect(experimentalAppendMove(new Alg("R U R'"), new Move("U2"))).to.equal(
+      new Alg("R U R' U"),
+    );
     expect(
       experimentalAppendMove(new Alg("R U R'"), new Move("R", -2)).isIdentical(
         new Alg("R U R' R2'"),
@@ -56,22 +58,26 @@ describe("operation", () => {
   it("computes mod offsets correctly", () => {
     expect(
       experimentalAppendMove(new Alg("L3"), new Move("L"), {
-        coalesce: true, mod: 4,
+        coalesce: true,
+        mod: 4,
       }).isIdentical(new Alg("")),
     ).to.be.true;
     expect(
       experimentalAppendMove(new Alg("L3"), new Move("L3"), {
-        coalesce: true, mod: 4,
+        coalesce: true,
+        mod: 4,
       }).isIdentical(new Alg("L2")),
     ).to.be.true;
     expect(
       experimentalAppendMove(new Alg("L3"), new Move("L6"), {
-        coalesce: true, mod: 4,
+        coalesce: true,
+        mod: 4,
       }).isIdentical(new Alg("L")),
     ).to.be.true;
     expect(
       experimentalAppendMove(new Alg("L"), new Move("L"), {
-        coalesce: true, mod: 3,
+        coalesce: true,
+        mod: 3,
       }).isIdentical(new Alg("L'")),
     ).to.be.true;
   });
diff --git a/src/test/chai-workaround/index.ts b/src/test/chai-workaround/index.ts
index c1c57bd2..4a6182e5 100644
--- a/src/test/chai-workaround/index.ts
+++ b/src/test/chai-workaround/index.ts
@@ -7,5 +7,8 @@
  * So we work around both issues here.
  */
 
-const { expect: untypedExpect } = await import("@esm-bundle" + "/chai");
+const { expect: untypedExpect, use: untypedUse } = await import(
+  "@esm-bundle" + "/chai"
+);
 export const expect: typeof import("chai").expect = untypedExpect;
+export const use: typeof import("chai").use = untypedUse;
```

Is this more or less what you had in mind?

**Comment by anicolao** [comment]
In case it's easier to see the above diff as a PR, here is a PR that's not intended for committing. https://github.com/cubing/cubing.js/pull/210

**Comment by anicolao** [comment]
I'm a bit blocked on this because setting up these tests so that I can write tests is in the way of completing a PR for https://github.com/cubing/cubing.js/issues/207 but of the 3 or 4 proposals discussed here none seem satisfactory. Can we pick one of

- my original option 1
- my original option 2
- overriding `.equals`
- do nothing and write new tests in the existing style

I've excluded adding `.identicalAlg` to Chai's assertions because I understand that approach the least of all the options and to my eye it adds no advantages. 

I don't really mind which one you want done, as I have little long term interest in how these tests look except that they're in my way for getting slice and wide moves supported for bluetooth input. I do continue to think that the existing tests are awkward to write, awkward to read, and difficult to debug when they fail, so it'd be nice to have some change, but I'll take no change + forward progress over the current state :-)

**Comment by lgarron** [comment]
> * do nothing and write new tests in the existing style

This certainly sounds fine to me, as long as the error messages work for you!

> I've excluded adding `.identicalAlg` to Chai's assertions because I understand that approach the least of all the options and to my eye it adds no advantages.

I'm concerned that overloading `equals` could mask bugs. If we use our own method, there's no chance of type coercion or other bugs that accidentally cause a test to pass. That might sound a bit overly paranoid in this specific situation, but after having to deal with similar issues in other places I don't want to risk it.

> but I'll take no change + forward progress over the current state :-)

It's a deal. 😃

**Comment by anicolao** [comment]
OK feel free to close this as wont-fix. I have written a script that generates tests in the old style or in my style, so I can easily debug failing tests by regenerating the test file so that the output has meaning, and then once the tests are passing I can regenerate the style that gets checked in. So I no longer have any need to fix the underlying code.

**Comment by anicolao** [comment]
Thank you for adding the new `identicalAlg` assert. See https://github.com/cubing/cubing.js/pull/213

**Comment by anicolao** [comment]
Now that's landed, I think this can be closed.
